arch/risc-v/src/common/riscv_swint.c

```diff
@@ -415,6 +429,24 @@ int riscv_swint(int irq, void *context, void *arg)
           regs[REG_A1]         = regs[REG_A2]; /* signal */
           regs[REG_A2]         = regs[REG_A3]; /* info */
           regs[REG_A3]         = regs[REG_A4]; /* ucontext */
+
+#ifdef CONFIG_ARCH_KERNEL_STACK
+          /* If we are signalling a user process, then we must be operating
+           * on the kernel stack now.  We need to switch back to the user
+           * stack before dispatching the signal handler to the user code.
+           * The existence of an allocated kernel stack is sufficient
+           * information to make this decision.
+           */
+
+          if (rtcb->xcp.kstack != NULL)
+            {
+              DEBUGASSERT(rtcb->xcp.kstkptr == NULL &&
+                          rtcb->xcp.ustkptr != NULL);
+
+              rtcb->xcp.kstkptr = (uintptr_t *)regs[REG_SP];
+              regs[REG_SP]      = (uintptr_t)rtcb->xcp.ustkptr;
```

Comment:
I think the assert test you mentioned below is to ensure that the stack pointer has been fully unwound to what was set here?